File: apzkr-pzpi-21-10-karasov-mykhailo/Task1-Server/src/core/common/uttils/JWT.ts
import * as jwt from 'jsonwebtoken';
import {JWT_SECRET_KEY} from "../../../config";
import UserDomainModel from "../../domain/models/User/User";
import RoleDomainModel from "../../domain/models/Role/Role";


export default class JWT {

    private readonly payload: any;

    private readonly _user: UserDomainModel;
    private readonly _roles: RoleDomainModel[];

    constructor(user: UserDomainModel, roles: RoleDomainModel[]) {
        this._user = user;
        this._roles = roles;
    }

    generateJwt() {

        const payload = {
            id: this._user.id,
            email: this._user.email,
            firstName: this._user.firstName,
            secondName: this._user.secondName,
            birthday: this._user.birthday,
            userImage: this._user.userImage,
            phoneNumber: this._user.phoneNumber,
            companyId: this._user.companyId,
            roles: this._roles
        };

        return jwt.sign(payload, JWT_SECRET_KEY, {
            expiresIn: '24h'
        })
    }
}

File: apzkr-pzpi-21-10-karasov-mykhailo/Task4-MobileClient/app/src/main/java/com/example/tasksyncmobileapp/util/classes/Jwt.kt
package com.example.tasksyncmobileapp.util.classes
import com.auth0.jwt.JWT
import com.auth0.jwt.interfaces.DecodedJWT
import com.example.tasksyncmobileapp.model.Role
import com.example.tasksyncmobileapp.model.User
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken


class Jwt {
    fun decodeJWT(token: String): User {
        val jwt: DecodedJWT = JWT.decode(token)

        val id = jwt.claims["id"]?.asInt() ?: throw RuntimeException("Missing 'id' claim in JWT")
        val email = jwt.claims["email"]?.asString() ?: throw RuntimeException("Missing 'email' claim in JWT")
        val firstName = jwt.claims["firstName"]?.asString() ?: throw RuntimeException("Missing 'firstName' claim in JWT")
        val secondName = jwt.claims["secondName"]?.asString() ?: throw RuntimeException("Missing 'secondName' claim in JWT")
        val userImage = jwt.claims["userImage"]?.asString() ?: throw RuntimeException("Missing 'userImage' claim in JWT")
        val birthday = jwt.claims["birthday"]?.asString() ?: throw RuntimeException("Missing 'birthday' claim in JWT")
        val phoneNumber = jwt.claims["phoneNumber"]?.asString() ?: ""
        val companyId = jwt.claims["companyId"]?.asInt() ?: -1

        val rolesJson = jwt.claims["roles"].toString()
        println(rolesJson)
        val gson = Gson()

        val type = object : TypeToken<List<Role>>() {}.type

        val roles: List<Role> = gson.fromJson(rolesJson, type)


        return User(id, email, firstName, secondName, userImage, birthday, phoneNumber, companyId, roles)
    }

}

